extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/filters/GracefulShutdownFilter.java

```diff
@@ -0,0 +1,68 @@
+package io.quarkus.vertx.http.runtime.filters;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.jboss.logging.Logger;
+
+import io.netty.handler.codec.http.HttpResponseStatus;
+import io.quarkus.runtime.shutdown.ShutdownListener;
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpServerRequest;
+
+public class GracefulShutdownFilter implements ShutdownListener, Handler<HttpServerRequest> {
+
+    private static Logger log = Logger.getLogger(GracefulShutdownFilter.class);
+
+    private volatile Handler<HttpServerRequest> next;
+    private volatile boolean running = true;
+    private final AtomicInteger currentRequestCount = new AtomicInteger();
+    private final AtomicReference<ShutdownNotification> notification = new AtomicReference<>();
+
+    private final Handler<Void> requestDoneHandler = new Handler<Void>() {
+        @Override
+        public void handle(Void event) {
+            int count = currentRequestCount.decrementAndGet();
+            if (!running) {
+                if (count == 0) {
+                    ShutdownNotification n = notification.get();
+                    if (n != null) {
+                        if (notification.compareAndSet(n, null)) {
+                            n.done();
+                            log.info("All HTTP requests complete");
+                        }
+                    }
+                }
+            }
+        }
+    };
+
+    @Override
+    public void handle(HttpServerRequest event) {
+        if (!running) {
+            event.response().setStatusCode(HttpResponseStatus.SERVICE_UNAVAILABLE.code()).end();
+            return;
+        }
+        currentRequestCount.incrementAndGet();
+        //todo: some way to do this without a wrapper solution
+        next.handle(new ShutdownRequestWrapper(event, requestDoneHandler));
```

Comment:
@vietj is there a better way to handle this?

Basically I need a reliable way to know when a request is done so I know how many requests are running (that can't be overridden in a downstream handler), and wrapping the request and response like this was the only way I could figure out to do it.

Reply:
on the HTTP Server Response, you can attach a https://vertx.io/docs/apidocs/io/vertx/core/http/HttpServerResponse.html#endHandler-io.vertx.core.Handler- to be notified when the response has been written.

Reply:
It would need to be coupled with the closeHandler when the connection is closed (abruptly).